test/unit/controllers/declaration/ProcedureCodesControllerSpec.scala

```diff
@@ -0,0 +1,257 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package unit.controllers.declaration
+
+import controllers.declaration.ProcedureCodesController
+import controllers.util.Remove
+import forms.Choice.AllowedChoiceValues.SupplementaryDec
+import forms.declaration.ProcedureCodes
+import models.declaration.ProcedureCodesData
+import models.declaration.ProcedureCodesData.limitOfCodes
+import org.mockito.ArgumentCaptor
+import org.mockito.ArgumentMatchers.any
+import org.mockito.Mockito.{reset, times, verify, when}
+import org.scalatest.OptionValues
+import play.api.data.Form
+import play.api.test.Helpers._
+import play.twirl.api.HtmlFormat
+import services.cache.ExportItem
+import unit.base.ControllerSpec
+import unit.mock.ErrorHandlerMocks
+import views.html.declaration.procedure_codes
+
+class ProcedureCodesControllerSpec extends ControllerSpec with ErrorHandlerMocks with OptionValues {
+
+  val mockProcedureCodesPage = mock[procedure_codes]
+
+  val controller = new ProcedureCodesController(
+    mockAuthAction,
+    mockJourneyAction,
+    mockErrorHandler,
+    mockExportsCacheService,
+    stubMessagesControllerComponents(),
+    mockProcedureCodesPage
+  )(ec)
+
+  val itemId = "itemId12345"
+
+  override protected def beforeEach(): Unit = {
+    super.beforeEach()
+    authorizedUser()
+    setupErrorHandler()
+    withNewCaching(aDeclaration(withChoice(SupplementaryDec)))
+    when(mockProcedureCodesPage.apply(any(), any(), any())(any(), any())).thenReturn(HtmlFormat.empty)
+  }
+
+  override protected def afterEach(): Unit = {
+    reset(mockProcedureCodesPage)
+    super.afterEach()
+  }
+
+  def theResponse: (Form[ProcedureCodes], Seq[String]) = {
+    val formCaptor = ArgumentCaptor.forClass(classOf[Form[ProcedureCodes]])
+    val dataCaptor = ArgumentCaptor.forClass(classOf[Seq[String]])
+    verify(mockProcedureCodesPage).apply(any(), formCaptor.capture(), dataCaptor.capture())(any(), any())
+    (formCaptor.getValue, dataCaptor.getValue)
+  }
+
+  "Procedure Codes controller" should {
+
+    "return 200 (OK)" when {
+
+      "display page method is invoked with empty cache" in {
+
+        val result = controller.displayPage(itemId)(getRequest())
+
+        status(result) mustBe OK
+        verify(mockProcedureCodesPage, times(1)).apply(any(), any(), any())(any(), any())
+
+        val (responseForm, responseSeq) = theResponse
+        responseForm.value mustBe empty
+        responseSeq mustBe empty
+      }
+
+      "display page method is invoked with data in cache" in {
+
+        val item = ExportItem(itemId, procedureCodes = Some(ProcedureCodesData(Some("1234"), Seq("123"))))
+        withNewCaching(aDeclaration(withItem(item)))
```

Comment:
here you make stub two times - one in `beforeEach` and second time here.

I think if something does not fit for each test it should not be in before each

Reply:
Fixed